storage/engine/mvcc.go

```diff
@@ -420,56 +422,59 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
+// most recent non-intent value instead. In the event that an inconsistent read
+// does encounter an intent (currently there can only be one), it is returned
+// via the proto.Intent slice, in addition to the result.
+func mvccGetInternal(engine Engine, key proto.Key, metaKey proto.EncodedKey,
+	timestamp proto.Timestamp, consistent bool, txn *proto.Transaction,
+	getValue getValueFunc, buf *getBuffer) (*proto.Value, []proto.Intent, error) {
 	if !consistent && txn != nil {
-		return nil, util.Errorf("cannot allow inconsistent reads within a transaction")
+		return nil, nil, util.Errorf("cannot allow inconsistent reads within a transaction")
 	}
 
-	var err error
 	meta := &buf.meta
 
 	// If value is inline, return immediately; txn & timestamp are irrelevant.
 	if meta.IsInline() {
 		if err := meta.Value.Verify(key); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
-		return meta.Value, nil
+		return meta.Value, nil, nil
 	}
-	// If we're doing inconsistent reads and there's an intent, we
-	// ignore the intent by insisting that the timestamp we're reading
-	// at is a historical timestamp < the intent timestamp. However, we
-	// return a write intent error so that the intents can be resolved.
-	var wiErr error
+	var ignoredIntents []proto.Intent
 	if !consistent && meta.Txn != nil && !timestamp.Less(meta.Timestamp) {
+		// If we're doing inconsistent reads and there's an intent, we
+		// ignore the intent by insisting that the timestamp we're reading
+		// at is a historical timestamp < the intent timestamp. However, we
+		// return the intent separately; the caller may want to resolve it.
+		ignoredIntents = append(ignoredIntents, proto.Intent{Key: key, Txn: *meta.Txn})
 		timestamp = meta.Timestamp.Prev()
-		wiErr = &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
 	}
 
 	var valueKey proto.EncodedKey
 	value := &buf.value
 
-	// First case: Our read timestamp is ahead of the latest write, or the
-	// latest write and current read are within the same transaction.
-	if !timestamp.Less(meta.Timestamp) ||
-		(meta.Txn != nil && txn != nil && bytes.Equal(meta.Txn.ID, txn.ID)) {
-		if meta.Txn != nil && (txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID)) {
-			// Trying to read the last value, but it's another transaction's
-			// intent; the reader will have to act on this.
-			return nil, &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
+	if !timestamp.Less(meta.Timestamp) && meta.HasWriteIntentError(txn) {
+		// Trying to read the last value, but it's another transaction's
+		// intent; the reader will have to act on this.
+		return nil, nil, &proto.WriteIntentError{
+			Intents: []proto.Intent{{Key: key, Txn: *meta.Txn}},
 		}
+	} else if ownIntent := meta.IsIntentOf(txn); !timestamp.Less(meta.Timestamp) || ownIntent {
+		// We are reading the latest value, which is not an intent or our intent
```

Comment:
"which is either an intent written by this transaction or not an intent at all (so there's no conflict)."